packages/react/focus-scope/src/FocusScope.tsx

```diff
@@ -26,44 +35,265 @@ type FocusScopeProps = {
   React.useEffect(() => {
-    const container = containerRef.current;
     if (container) {
-      container.addEventListener(AUTOFOCUS_ON_CREATE, onMountAutoFocus);
-      container.addEventListener(AUTOFOCUS_ON_DESTROY, onUnmountAutoFocus);
-      focusScopeRef.current = createFocusScope(container);
+      focusScopesStack.add(scopeRef);
+      const PREVIOUSLY_FOCUSED_ELEMENT = document.activeElement as HTMLElement | null;
+
+      // we need to setup the listeners before we `dispatchEvent`
+      container.addEventListener(AUTOFOCUS_ON_MOUNT, onMountAutoFocus);
+      container.addEventListener(AUTOFOCUS_ON_UNMOUNT, onUnmountAutoFocus);
+
+      const mountEvent = new Event(AUTOFOCUS_ON_MOUNT, EVENT_OPTIONS);
+      container.dispatchEvent(mountEvent);
+      if (!mountEvent.defaultPrevented) {
+        focusFirst(getTabbableCandidates(container), { select: true });
+        if (document.activeElement === PREVIOUSLY_FOCUSED_ELEMENT) {
+          focus(container);
+        }
+      }
 
       return () => {
-        container.removeEventListener(AUTOFOCUS_ON_CREATE, onMountAutoFocus);
+        container.removeEventListener(AUTOFOCUS_ON_MOUNT, onMountAutoFocus);
+
         // We hit a react bug (fixed in v17) with focusing in unmount.
         // We need to delay the focus a little to get around it for now.
         // See: https://github.com/facebook/react/issues/17894
         setTimeout(() => {
-          focusScopeRef.current?.destroy();
-          // this needs to come after calling `destroy()` to make sure we can catch the event on time.
-          container.removeEventListener(AUTOFOCUS_ON_DESTROY, onUnmountAutoFocus);
+          const unmountEvent = new Event(AUTOFOCUS_ON_UNMOUNT, EVENT_OPTIONS);
+          container.dispatchEvent(unmountEvent);
+          if (!unmountEvent.defaultPrevented) {
+            focus(PREVIOUSLY_FOCUSED_ELEMENT ?? document.body, { select: true });
+          }
+          // we need to remove the listener after we `dispatchEvent`
+          container.removeEventListener(AUTOFOCUS_ON_UNMOUNT, onUnmountAutoFocus);
+
+          focusScopesStack.remove(scopeRef);
         }, 0);
       };
     }
-  }, [containerRef, onMountAutoFocus, onUnmountAutoFocus]);
+  }, [container, onMountAutoFocus, onUnmountAutoFocus]);
+
+  // Takes care of containing focus and wrapping focus (when tabbing whilst at the edges)
+  const handleKeyDown = React.useCallback(
+    (event: React.KeyboardEvent) => {
+      if (!wrapped && !contained) return;
+      if (scopeRef.current.paused) return;
+
+      if (event.key === 'Tab' && !event.altKey && !event.ctrlKey && !event.metaKey) {
+        const focusedElement = document.activeElement as HTMLElement | null;
+        if (focusedElement) {
+          const container = event.currentTarget as HTMLElement;
+          const [first, last] = getTabbableEdges(container);
+          if (!first || !last) {
+            event.preventDefault();
```

Comment:
am i understanding this correctly that if there is no first or last element then we don't let them tab? if so, why not? is there an implicit assumption here that the `focusedElement` is the container?

Reply:
Yeah, if there's not first and last, then it means nothing is tabbable so tabbing shouldn't do anything. It's already covered by the containment, but in the case of just wrapping, then I think something like this is needed anyway. But I agree it feels a bit implicit right now. I'll see what I can do.